### PR TITLE
Fix QPY serialisation of ControlledGate with open controls

### LINKED ISSUE #8549: qpy changes names of `ControlledGate`s with non-default `ctrl_state`
### Environment

- **Qiskit Terra version**: 0.22.1
- **Python version**: 3.8
- **Operating system**: Ubuntu 22.04.1

### What is happening?

qpy appends an extra `"_o{ctrl_state}"` to the name of `qiskit.circuit.ControlledGate` instances with non-default values for `ctrl_state` (so after one serialization+deserialization round the `gate.name` becomes `"{name}_o{ctrl_state}_o{ctrl_state}"`, and after a second round becomes `"{name}_o{ctrl_state}_o{ctrl_state}_o{ctrl_state}"`, etc)

### How can we reproduce the issue?

```
import qiskit
import qiskit.qpy
import io

gate = qiskit.circuit.library.SGate().control(1, ctrl_state="0")  # create an instance of ControlledGate
print(gate.name)  # prints "cs_o0"

circuit = qiskit.circuit.QuantumCircuit(2)
circuit.append(gate, [0, 1])
print(circuit[0].operation.name)  # prints "cs_o0"

qiskit.qpy.dump(circuit, buf := io.BytesIO())
circuit = qiskit.qpy.load(io.BytesIO(buf.getbuffer()))[0]
print(circuit[0].operation.name)  # prints "cs_o0_o0"

qiskit.qpy.dump(circuit, buf := io.BytesIO())
circuit = qiskit.qpy.load(io.BytesIO(buf.getbuffer()))[0]
print(circuit[0].operation.name)  # prints "cs_o0_o0_o0"
```

### What should happen?

every `print` statement above should print `"cs_o0"`

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

Previously, an incorrect definition and name would be re-instated on
QPY deserialisation of a `ControlledGate` instance with open controls.
The base name would include the dynamic `_o{ctrl_state}` suffix, causing
the suffix to later be duplicated, and the definition would duplicate
the logic that added the open controls.  This fixes both by stripping
the suffix on re-read before it is assigned, and serialising only the
"11...1" state definition, since this is what is required and stored by
`ControlledGate`.

### Details and comments

Fix #8549.

This is technically a breaking change in the behaviour of QPY (it changes how something is serialised and deserialised), but it is not a format change.  A file created with the previous serialisation behaviour will still load with this new logic, and an older Qiskit will be able to deserialise a circuit dumped with this logic, but in both cases, a `ControlledGate` with open controls will not be restored to the expected name/definition.  There will be no exception, it'll just produce the wrong result, as it did before this patch.

### CODE REVIEW

**Review by mtreinish — COMMENTED** [review]
LGTM, do you want to add a backwards compat test for this case too? I'm thinking it might be a good idea since the upgrade path on this seems a bit nuanced so we probably don't want it to regress moving forward.

File: test/qpy_compat/test_qpy.py [hunk 1]
```diff
@@ -525,6 +549,8 @@ def generate_circuits(version_str=None):
         output_circuits["controlled_gates.qpy"] = generate_controlled_gates()
         output_circuits["schedule_blocks.qpy"] = generate_schedule_blocks()
         output_circuits["pulse_gates.qpy"] = generate_calibrated_circuits()
+    if version_parts >= (0, 21, 2):
```
Comment (jakelishman) [inline test/qpy_compat/test_qpy.py:552]: I've made this 0.21.2 assuming we might throw this in the backports, but if we're leaving it for 0.22.0, feel free to just modify this line.

**Review by jakelishman — COMMENTED** [review]
Yeah, the backwards compatibility test sounds good - added in this commit.  There's already a test of `ctrl_state=0`, but unfortunately it's on a CX gate, which doesn't have the same issues because it's a known gate.

File: test/qpy_compat/test_qpy.py (hunk 1 above)
Comment (mtreinish) [reply test/qpy_compat/test_qpy.py:552]: Yeah, that's fine I probably would have just done `> (0.21.1)` to cover either case but this works fine as I was intending to tag this as stable backport potential

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for adding the backwards compat test

**Pull request merged**